_posts/2020-11-09-defensive-programing.md

```diff
@@ -0,0 +1,110 @@
+        throw new ArgumentException("Date is too early");
+
+    if (dateTime.Date > DateTime.Now.AddMonths(1).Date)
+        throw new ArgumentException("Date is too late");
+
+    /* Create an appointment */
+}
+```
+Here, the code will only make the appointment if the the given date is correct, not too early or too late.
+
+* **Nulls**
+
+Null is a pointer used in programming for uninitialized, undefined, empty, or meaningless value. Because many programs cannot discriminate null values, they are often the reason for bugs in codes. Therefore, checking your code for nulls when applicable is always a good idea.
+
+Example:
+```r
+public class Controller
+{
+    public Controller(ILogger logger, IEmailGateway gateway)
+    {
+        if (logger == null)
+            throw new ArgumentNullException();
+        if (gateway == null)
+            throw new ArgumentNullException();
+
+        /* ... */
+    }
+
+    public void Process(User user, Order order)
+    {
+        if (user == null)
+            throw new ArgumentNullException();
+
+        /* ... */
+    }
+}
+```
+In this case, the code is checking for `null` values.
+
+## Why is defensive programming important?
+
+Defensive programming is a good practice that allows the user to react to unexpected errors and fix them faster. Let’s say that you create a code that processes extensive tabular data, but you do not use defensive programming. The code works perfectly with one dataset, great! However, when you try to use the same code to process a similar tabular dataset.  
+In this case, you first have to find where in the code did the error happen, and then you have to figure out why it happened. Perhaps the argument that a function is looking for does not exist in the new input file, or it exists but it is not in the correct format; any of these errors can take you two minutes to fix, but hours to find.  
+Troubleshooting can take a lot of time, and it is probably part of every programmer’s worst nightmares. Now, by using defensive programming style, the guard statements and assertions can guide you in the direction of the problem, or even tell you exactly what the problem was, saving time and protecting your mental health.  
+
+## Final thoughts  
+
+"Defensive programming is good programming". This phrase caught my attention the first time I heard it, and now I think it might be one of the most important things I have learned about programming. When writing a code that will later be used with different input data or by someone else, it is important to think about how you can make things easier for the user. Defensive programming helps programmers write codes that are prepared to deal with any unexpected circumstances, and even provide guidance on how to solve the issue. It helps to create self-checking codes that can be widely used, and I believe that is a beautiful thing. So, let's use defensive programming and make the coding world a better place!
```

Comment:
Nice summary and I like that you included additional links with more information on the topic!